File: benchmarks/v10_external/trajectory.py
"""Read the frozen HOLOSTEM1 trajectory and build its fixed Rips complex."""

from __future__ import annotations

import math
from dataclasses import dataclass, replace
from itertools import pairwise
from pathlib import Path


@dataclass(frozen=True)
class Edge:
    """One canonical edge in the fixed graph."""

    u: int
    v: int


@dataclass(frozen=True)
class Trajectory:
    """A validated sequence of edge-weight snapshots."""

    dataset: str
    source_sha256: str
    vertices: int
    edges: tuple[Edge, ...]
    snapshot_times: tuple[int, ...]
    weights: tuple[tuple[float, ...], ...]


@dataclass(frozen=True)
class _Header:
    """Validated trajectory fields needed to decode snapshots."""

    dataset: str
    source_sha256: str
    vertices: int
    edge_count: int
    snapshot_count: int


@dataclass(frozen=True)
class Complex:
    """The fixed vertices, edges, and induced triangles used by both arms."""

    simplices: tuple[tuple[int, ...], ...]
    dimensions: tuple[int, ...]
    edge_positions: dict[tuple[int, int], int]


class _Tokens:
    def __init__(self, path: Path) -> None:
        self.values = path.read_text(encoding="ascii").split()
        self.cursor = 0

    def take(self, label: str) -> str:
        if self.cursor >= len(self.values):
            raise ValueError(f"missing {label}")
        value = self.values[self.cursor]
        self.cursor += 1
        return value

    def expect(self, label: str) -> None:
        actual = self.take(label)
        if actual != label:
            raise ValueError(f"expected {label}, found {actual}")

    def integer(self, label: str) -> int:
        value = self.take(label)
        try:
            parsed = int(value, 10)
        except ValueError as error:
            raise ValueError(f"invalid {label}: {value}") from error
        if parsed < 0:
            raise ValueError(f"invalid {label}: {value}")
        return parsed

    def real(self, label: str) -> float:
        value = self.take(label)
        try:
            parsed = float(value)
        except ValueError as error:
            raise ValueError(f"invalid {label}: {value}") from error
        if not math.isfinite(parsed) or parsed < 0:
            raise ValueError(f"invalid {label}: {value}")
        return parsed


def read_trajectory(path: Path) -> Trajectory:
    """Read and validate one deterministic HOLOSTEM1 trajectory."""

    tokens = _Tokens(path)
    header = _read_header(tokens)
    snapshot_times = _read_snapshot_times(tokens, header.snapshot_count)
    edges, weights = _read_edges(tokens, header)
    if tokens.cursor != len(tokens.values):
        raise ValueError("trajectory has trailing fields")
    return Trajectory(
        header.dataset,
        header.source_sha256,
        header.vertices,
        tuple(edges),
        snapshot_times,
        tuple(tuple(snapshot) for snapshot in weights),
    )


def _read_header(tokens: _Tokens) -> _Header:
    tokens.expect("HOLOSTEM1")
    tokens.expect("dataset")
    dataset = tokens.take("dataset name")
    tokens.expect("source_sha256")
    source_sha256 = tokens.take("source SHA-256")
    if not _is_sha256(source_sha256):
        raise ValueError("source SHA-256 is not 64 hexadecimal digits")

    vertices, edge_count, snapshot_count = _read_counts(tokens)
    _validate_counts(vertices, edge_count, snapshot_count)
    _read_preprocessing(tokens)
    return _Header(dataset, source_sha256, vertices, edge_count, snapshot_count)


def _is_sha256(value: str) -> bool:
    return len(value) == 64 and all(
        character in "0123456789abcdefABCDEF" for character in value
    )


def _read_counts(tokens: _Tokens) -> tuple[int, int, int]:
    tokens.expect("vertices")
    vertices = tokens.integer("vertex count")
    tokens.expect("edges")
    edge_count = tokens.integer("edge count")
    tokens.expect("snapshots")
    snapshot_count = tokens.integer("snapshot count")
    return vertices, edge_count, snapshot_count


def _validate_counts(vertices: int, edge_count: int, snapshot_count: int) -> None:
    if vertices == 0 or edge_count == 0 or snapshot_count < 2:
        raise ValueError("trajectory counts are outside the benchmark domain")
    if edge_count > 20_000_000 or snapshot_count > 1_000_000:
        raise ValueError("trajectory exceeds the benchmark size limit")
    if edge_count > 20_000_000 // snapshot_count:
        raise ValueError("trajectory exceeds the benchmark cell limit")


def _read_preprocessing(tokens: _Tokens) -> None:

    tokens.expect("bin_seconds")
    if tokens.integer("bin_seconds") == 0:
        raise ValueError("bin_seconds must be positive")
    tokens.expect("warmup_bins")
    if tokens.integer("warmup_bins") == 0:
        raise ValueError("warmup_bins must be positive")
    tokens.expect("decay_numerator")
    decay_numerator = tokens.integer("decay_numerator")
    tokens.expect("decay_denominator")
    decay_denominator = tokens.integer("decay_denominator")
    if decay_denominator == 0 or decay_numerator >= decay_denominator:
        raise ValueError("invalid decay constants")
    tokens.expect("score_scale")
    if tokens.integer("score_scale") == 0:
        raise ValueError("score_scale must be positive")
    tokens.expect("weight_offset")
    weight_offset = tokens.integer("weight_offset")
    if weight_offset == 0:
        raise ValueError("weight_offset must be positive")
    tokens.expect("maximum_score")
    maximum_score = tokens.integer("maximum_score")
    if maximum_score >= weight_offset:
        raise ValueError("maximum_score must be below weight_offset")


def _read_snapshot_times(tokens: _Tokens, snapshot_count: int) -> tuple[int, ...]:
    tokens.expect("snapshot_times")
    snapshot_times = tuple(
        tokens.integer("snapshot time") for _ in range(snapshot_count)
    )
    if any(left >= right for left, right in pairwise(snapshot_times)):
        raise ValueError("snapshot times are not strictly increasing")
    return snapshot_times


def _read_edges(
    tokens: _Tokens, header: _Header
) -> tuple[list[Edge], list[list[float]]]:
    tokens.expect("edge_weights")
    edges: list[Edge] = []
    snapshots = [[] for _ in range(header.snapshot_count)]
    previous: Edge | None = None
    for _ in range(header.edge_count):
        edge = _read_edge(tokens, header.vertices, previous)
        edges.append(edge)
        previous = edge
        _read_edge_weights(tokens, snapshots, edge)
    return edges, snapshots


def _read_edge(tokens: _Tokens, vertices: int, previous: Edge | None) -> Edge:
    edge = Edge(tokens.integer("edge endpoint"), tokens.integer("edge endpoint"))
    if (
        edge.u >= edge.v
        or edge.v >= vertices
        or (previous is not None and (edge.u, edge.v) <= (previous.u, previous.v))
    ):
        raise ValueError("trajectory edges are not canonical")
    return edge


def _read_edge_weights(
    tokens: _Tokens, snapshots: list[list[float]], edge: Edge
) -> None:
    for snapshot in snapshots:
        snapshot.append(tokens.real("edge weight"))


def build_complex(trajectory: Trajectory) -> Complex:
    """Build every induced triangle of the fixed listed graph."""

    simplices, edge_positions, adjacency = _vertices_and_edges(trajectory)
    simplices.extend(_triangles(trajectory.vertices, adjacency))
    dimensions = tuple(len(simplex) - 1 for simplex in simplices)
    return Complex(tuple(simplices), dimensions, edge_positions)


def _vertices_and_edges(
    trajectory: Trajectory,
) -> tuple[list[tuple[int, ...]], dict[tuple[int, int], int], list[list[bool]]]:
    simplices = [(vertex,) for vertex in range(trajectory.vertices)]
    edge_positions: dict[tuple[int, int], int] = {}
    adjacency = [[False] * trajectory.vertices for _ in range(trajectory.vertices)]
    for position, edge in enumerate(trajectory.edges):
        edge_positions[(edge.u, edge.v)] = position
        simplices.append((edge.u, edge.v))
        adjacency[edge.u][edge.v] = True
        adjacency[edge.v][edge.u] = True
    return simplices, edge_positions, adjacency


def _triangles(
    vertices: int, adjacency: list[list[bool]]
) -> list[tuple[int, int, int]]:
    triangles: list[tuple[int, int, int]] = []
    for u in range(vertices):
        for v in range(u + 1, vertices):
            if not adjacency[u][v]:
                continue
            for w in range(v + 1, vertices):
                if adjacency[u][w] and adjacency[v][w]:
                    triangles.append((u, v, w))
    return triangles


def global_rank_trajectory(trajectory: Trajectory) -> Trajectory:
    """Encode all edge weights by exact global ranks for Dionysus' f32 data type."""

    distinct = sorted(
        {weight for snapshot in trajectory.weights for weight in snapshot}
    )
    if len(distinct) >= 2**24:
        raise ValueError("Dionysus rank encoding exceeds the exact f32 integer range")
    ranks = {weight: float(index + 1) for index, weight in enumerate(distinct)}
    weights = tuple(
        tuple(ranks[weight] for weight in snapshot) for snapshot in trajectory.weights
    )
    return replace(trajectory, weights=weights)


def snapshot_values(complex_: Complex, snapshot: tuple[float, ...]) -> list[float]:
    """Return one filtration value for each fixed simplex."""

    values: list[float] = []
    for simplex in complex_.simplices:
        if len(simplex) == 1:
            values.append(0.0)
            continue
        diameter = 0.0
        for left, u in enumerate(simplex):
            for v in simplex[left + 1 :]:
                try:
                    edge_position = complex_.edge_positions[(u, v)]
                except KeyError as error:
                    raise ValueError(
                        f"complex contains missing edge {(u, v)}"
                    ) from error
                diameter = max(diameter, snapshot[edge_position])
        values.append(diameter)
    return values

File: benchmarks/v10_external/trajectory.cpp
#include "trajectory.hpp"

#include <algorithm>
#include <cmath>
#include <cctype>
#include <fstream>
#include <limits>
#include <stdexcept>
#include <string>
#include <utility>

namespace holos_external {
namespace {

class Tokens {
public:
    explicit Tokens(const std::string& path) : input_(path) {
        if (!input_) {
            throw std::runtime_error("cannot open trajectory: " + path);
        }
    }

    std::string next(const char* label) {
        std::string value;
        if (!(input_ >> value)) {
            throw std::runtime_error(std::string("missing ") + label);
        }
        return value;
    }

    void expect(const char* expected) {
        const auto actual = next(expected);
        if (actual != expected) {
            throw std::runtime_error(
                "expected " + std::string(expected) + ", found " + actual);
        }
    }

    std::size_t size(const char* label) {
        const auto value = next(label);
        std::size_t consumed = 0;
        unsigned long long parsed = 0;
        try {
            parsed = std::stoull(value, &consumed, 10);
        } catch (const std::exception&) {
            throw std::runtime_error(std::string("invalid ") + label + ": " + value);
        }
        if (consumed != value.size() || parsed > std::numeric_limits<std::size_t>::max()) {
            throw std::runtime_error(std::string("invalid ") + label + ": " + value);
        }
        return static_cast<std::size_t>(parsed);
    }

    double real(const char* label) {
        const auto value = next(label);
        std::size_t consumed = 0;
        double parsed = 0.0;
        try {
            parsed = std::stod(value, &consumed);
        } catch (const std::exception&) {
            throw std::runtime_error(std::string("invalid ") + label + ": " + value);
        }
        if (consumed != value.size() || !std::isfinite(parsed) || parsed < 0.0) {
            throw std::runtime_error(std::string("invalid ") + label + ": " + value);
        }
        return parsed;
    }

    bool has_more() {
        std::string value;
        return static_cast<bool>(input_ >> value);
    }

private:
    std::ifstream input_;
};

void expect_positive(const char* label, std::size_t value) {
    if (value == 0) {
        throw std::runtime_error(std::string(label) + " must be positive");
    }
}

struct Header {
    std::string dataset;
    std::string source_sha256;
    std::size_t vertices;
    std::size_t edge_count;
    std::size_t snapshot_count;
};

std::string read_source_digest(Tokens& tokens) {
    tokens.expect("source_sha256");
    const auto digest = tokens.next("source SHA-256");
    const auto hexadecimal = std::all_of(
        digest.begin(), digest.end(),
        [](unsigned char value) { return std::isxdigit(value) != 0; });
    if (digest.size() != 64 || !hexadecimal) {
        throw std::runtime_error("source SHA-256 is not 64 hexadecimal digits");
    }
    return digest;
}

Header read_header(Tokens& tokens) {
    tokens.expect("HOLOSTEM1");
    tokens.expect("dataset");
    const auto dataset = tokens.next("dataset name");
    const auto source_sha256 = read_source_digest(tokens);
    tokens.expect("vertices");
    const auto vertices = tokens.size("vertex count");
    tokens.expect("edges");
    const auto edge_count = tokens.size("edge count");
    tokens.expect("snapshots");
    const auto snapshot_count = tokens.size("snapshot count");
    return {dataset, source_sha256, vertices, edge_count, snapshot_count};
}

void validate_counts(const Header& header) {
    expect_positive("vertex count", header.vertices);
    expect_positive("edge count", header.edge_count);
    if (header.snapshot_count < 2) {
        throw std::runtime_error("snapshot count must be at least two");
    }
    if (header.edge_count > 20'000'000 || header.snapshot_count > 1'000'000 ||
        header.edge_count > 20'000'000 / header.snapshot_count) {
        throw std::runtime_error("trajectory exceeds the benchmark cell limit");
    }
}

void read_constants(Tokens& tokens) {
    tokens.expect("bin_seconds");
    expect_positive("bin_seconds", tokens.size("bin_seconds"));
    tokens.expect("warmup_bins");
    expect_positive("warmup_bins", tokens.size("warmup_bins"));
    tokens.expect("decay_numerator");
    const auto decay_numerator = tokens.size("decay_numerator");
    tokens.expect("decay_denominator");
    const auto decay_denominator = tokens.size("decay_denominator");
    if (decay_denominator == 0 || decay_numerator >= decay_denominator) {
        throw std::runtime_error("invalid decay constants");
    }
    tokens.expect("score_scale");
    expect_positive("score_scale", tokens.size("score_scale"));
    tokens.expect("weight_offset");
    const auto weight_offset = tokens.size("weight_offset");
    expect_positive("weight_offset", weight_offset);
    tokens.expect("maximum_score");
    if (tokens.size("maximum_score") >= weight_offset) {
        throw std::runtime_error("invalid maximum score");
    }
}

void read_snapshot_times(Tokens& tokens, std::size_t snapshot_count) {
    tokens.expect("snapshot_times");
    std::size_t previous_time = 0;
    for (std::size_t index = 0; index < snapshot_count; ++index) {
        const auto current = tokens.size("snapshot time");
        if (index != 0 && current <= previous_time) {
            throw std::runtime_error("snapshot times are not strictly increasing");
        }
        previous_time = current;
    }
}

void validate_edge(
    const std::vector<Edge>& edges,
    std::size_t vertices,
    std::size_t u,
    std::size_t v) {
    const auto out_of_order = !edges.empty() &&
                              (u < edges.back().u ||
                               (u == edges.back().u && v <= edges.back().v));
    if (u >= v || v >= vertices || out_of_order) {
        throw std::runtime_error("trajectory edges are not canonical");
    }
}

std::pair<std::vector<Edge>, std::vector<Snapshot>> read_weights(
    Tokens& tokens,
    const Header& header) {
    tokens.expect("edge_weights");
    std::vector<Edge> edges;
    edges.reserve(header.edge_count);
    std::vector<Snapshot> snapshots(header.snapshot_count);
    for (auto& snapshot : snapshots) {
        snapshot.weights.reserve(header.edge_count);
    }
    for (std::size_t edge_index = 0; edge_index < header.edge_count; ++edge_index) {
        const auto u = tokens.size("edge endpoint");
        const auto v = tokens.size("edge endpoint");
        validate_edge(edges, header.vertices, u, v);
        edges.push_back({u, v});
        for (auto& snapshot : snapshots) {
            snapshot.weights.push_back(tokens.real("edge weight"));
        }
    }
    return {std::move(edges), std::move(snapshots)};
}

}

Trajectory read_trajectory(const std::string& path) {
    Tokens tokens(path);
    const auto header = read_header(tokens);
    validate_counts(header);
    read_constants(tokens);
    read_snapshot_times(tokens, header.snapshot_count);
    auto [edges, snapshots] = read_weights(tokens, header);
    if (tokens.has_more()) {
        throw std::runtime_error("trajectory has trailing fields");
    }
    return Trajectory{
        header.dataset,
        header.source_sha256,
        header.vertices,
        std::move(edges),
        std::move(snapshots)};
}

}
